File: 3186-maximum-total-damage-with-spell-casting/3186-maximum-total-damage-with-spell-casting.cpp
class Solution {
public:
    long long maximumTotalDamage(vector<int>& power) {
        unordered_map<int, int> mp;
        for (int& x : power)
            mp[x]++;
        vector<int> value;
        for (auto& it : mp)
            value.push_back(it.first);
        sort(value.begin(), value.end());
        int n = value.size();
        vector<long long> dp(n);
        dp[0] = (long long)value[0] * mp[value[0]];
        for (int i = 1; i < n; i++) {
            long long take = value[i] * mp[value[i]];
            int prev = upper_bound(value.begin(), value.end(), value[i] - 3) -
                       value.begin() - 1;
            if (prev >= 0) {
                take += dp[prev];
            }
            dp[i] = max(dp[i - 1], take);
        }
        return dp[n - 1];
    }
};

File: 3186-maximum-total-damage-with-spell-casting/3186-maximum-total-damage-with-spell-casting.py
class Solution:
    def maximumTotalDamage(self, power: List[int]) -> int:
        freq = Counter(power) #like unordered_map in C++
        value = sorted(freq.keys())
        n = len(value)
        dp = [0] * n
        dp[0] = value[0] * freq[value[0]]
        for i in range(n):
            take = value[i] * freq[value[i]]
            prev = bisect.bisect_right(value, value[i] - 3) - 1   #like upper_bound in C++
            if prev >= 0:
                take += dp[prev]
            dp[i] = max(take, dp[i - 1])
        return dp[n - 1]
